haskell/src/Stablecoin/Client/Impl.hs

```diff
@@ -327,12 +316,57 @@ getMintingAllowance contract minter = do
 
--- | Get the token metadata of the contract
 getTokenMetadata :: "contract" :! AddressOrAlias -> MorleyClientM FA2.TokenMetadata
 getTokenMetadata contract = do
-  mdRegistry <- sTokenMetadataRegistry <$> getStorage contract
-  bigMapId <- mrsTokenMetadata <$> getRegistryStorage mdRegistry
-  readBigMapValue bigMapId FA2.theTokenId
+  (metadata, storageView) <- getContractMetadata contract
+  case TZ.getViews metadata of
+    Left err -> throwM $ SCEMetadataError ("Views was not found in metadata:" <> show err)
+    Right views -> case MD.findView @(ToT Storage) views "GetTokenMetadata" of
+      Nothing -> throwM $ SCEMetadataError "'GetTokenMetadata' view was not found in metadata"
+      Just getTMD -> do
+        let storageWithEmptyBm :: Storage =
+              -- In the below code we convert the `StorageView` to `Storage` by
+              -- replacing all bigmaps with empty counterparts, and this should be fine here since
+              -- the token metadata view does not access any of these big maps.
+              storageView
+                { sLedger = mempty
+                , sMetadata = mempty
+                , sOperators = mempty
+                , sPermits = mempty
+                }
+        case MD.interpretView @(Natural, Map MText ByteString)
+            dummyContractEnv getTMD (MD.ViewParam (0 :: Natural)) storageWithEmptyBm of
+          Right (_, md) -> pure md
+          Left err -> throwM $ SCEMetadataError ("Error while interpreting the contract:" <> show err)
+
+-- | Get the metadata of the contract
+getContractMetadata
+  :: "contract" :! AddressOrAlias
+  -> MorleyClientM (TZ.Metadata (ToT Storage), StorageView)
+getContractMetadata contract = do
+  storageView <- getStorage contract
+  let bigMapId = sMetadata storageView
+  metadataUri <- decodeUtf8 <$> readBigMapValue bigMapId [mt||]
+  case parseMetadataUri metadataUri of
+    Nothing -> throwM $ SCEMetadataError ("Unparsable URI:" <> metadataUri)
+    Just uri -> case uri of
+      InCurrentContractUnderKey key -> case mkMText key of
+        Right mtKey -> do
+          rawMd <- readBigMapValue bigMapId mtKey
+          case J.eitherDecodeStrict rawMd of
+            Right a -> pure (a, storageView)
+            Left err -> throwM $ SCEMetadataError ("Error decoding metadata:" <> show err)
+        Left _ -> throwM $ SCEMetadataError ("Unexpected key:" <> key)
+      InRemoteContractUnderKey addr key -> do
+        case mkMText key of
+          Right mtKey -> do
+            bigMapId_ <- snd <$> getMetadataRegistryStorage (#contract .! (AddressResolved addr))
+            rawMd <- readBigMapValue bigMapId_ mtKey
+            case J.eitherDecodeStrict rawMd of
+              Right a -> pure (a, storageView)
+              Left err -> throwM $ SCEMetadataError ("Error decoding metadata:" <> show err)
+          Left _ -> throwM $ SCEMetadataError ("Unexpected key:" <> key)
+      RawUri uri_ -> throwM $ SCEMetadataError ("Unsupported metadata URI:" <> uri_)
```

Comment:
It seems to me that here `throwM SCEMetadataError` is used for each `Left` case (plus a `Nothing` and the last lines), I suggest reducing the repetition and indentation of this code block by making a helper function like:
```haskell
throwMdErr toMsg f = either (throwM . SCEMetadataError . toMsg) pure =<< f 
```

Reply:
I have tried to do this.